nix/shell.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2022 Jetpack Technologies Inc and contributors. All rights reserved.
+// Use of this source code is governed by the license in the LICENSE file.
+
+package nix
+
+import (
+	"bytes"
+	_ "embed"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"text/template"
+
+	"github.com/pkg/errors"
+	"go.jetpack.io/devbox/debug"
+)
+
+//go:embed shellrc.tmpl
+var shellrcText string
+var shellrcTmpl = template.Must(template.New("shellrc").Parse(shellrcText))
+
+type name string
+
+const (
+	shUnknown name = ""
+	shBash    name = "bash"
+	shZsh     name = "zsh"
+	shKsh     name = "ksh"
+	shPosix   name = "posix"
+)
+
+// Shell configures a user's shell to run in Devbox. Its zero value is a
+// fallback shell that launches a regular Nix shell.
+type Shell struct {
+	name            name
+	binPath         string
+	userShellrcPath string
+
+	// UserInitHook contains commands that will run at shell startup.
+	UserInitHook string
+}
+
+// DetectShell attempts to determine the user's default shell.
+func DetectShell() (*Shell, error) {
+	path := os.Getenv("SHELL")
+	if path == "" {
+		return nil, errors.New("unable to detect the current shell")
+	}
+
+	sh := &Shell{binPath: filepath.Clean(path)}
+	base := filepath.Base(path)
+	// Login shell
+	if base[0] == '-' {
+		base = base[1:]
+	}
+	switch base {
+	case "bash":
+		sh.name = shBash
+		sh.userShellrcPath = rcfilePath(".bashrc")
+	case "zsh":
+		sh.name = shZsh
+		sh.userShellrcPath = rcfilePath(".zshrc")
+	case "ksh":
+		sh.name = shKsh
+		sh.userShellrcPath = rcfilePath(".kshrc")
+	case "dash", "ash", "sh":
+		sh.name = shPosix
+		sh.userShellrcPath = os.Getenv("ENV")
+
+		// Just make up a name if there isn't already an init file set
+		// so we have somewhere to put a new one.
+		if sh.userShellrcPath == "" {
+			sh.userShellrcPath = ".shinit"
+		}
+	default:
+		sh.name = shUnknown
+	}
+	debug.Log("Detected shell: %s", sh.binPath)
+	debug.Log("Recognized shell as: %s", sh.binPath)
+	debug.Log("Looking for user's shell init file at: %s", sh.userShellrcPath)
+	return sh, nil
+}
+
+// rcfilePath returns the absolute path for an rcfile, which is usually in the
+// user's home directory. It doesn't guarantee that the file exists.
+func rcfilePath(basename string) string {
+	home, err := os.UserHomeDir()
+	if err != nil {
+		return ""
+	}
+	return filepath.Join(home, basename)
+}
+
+func (s *Shell) Run(nixPath string) error {
+	// Launch a fallback shell if we couldn't find the path to the user's
+	// default shell.
+	if s.binPath == "" {
+		cmd := exec.Command("nix-shell", nixPath)
+		cmd.Stdin = os.Stdin
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+
+		debug.Log("Unrecognized user shell, falling back to: %v", cmd.Args)
```

Comment:
not sure: 
should the argument be `cmd.Args` - or `nix-shell nixPath` ?

Reply:
They should be roughly the same. This will print something like `["nix-shell" "/blah/.devbox/gen"]`. I wanted to use `cmd.Args` to guarantee we're printing exactly what's being launched.